communication/src/coap_channel_new.cpp

```diff
@@ -0,0 +1,1457 @@
+}
+
+void CoapChannel::destroyMessage(coap_message* apiMsg) {
+    if (!apiMsg) {
+        return;
+    }
+    auto msg = RefCountPtr<CoapMessage>::wrap(reinterpret_cast<CoapMessage*>(apiMsg)); // Take ownership
+    clearMessage(msg);
+}
+
+void CoapChannel::cancelRequest(int requestId) {
+    if (requestId <= 0) {
+        return;
+    }
+    // Search among the messages for which a user callback may still need to be invoked
+    RefCountPtr<CoapMessage> msg = findRefInList(sentReqs_, [=](auto req) {
+        return req->type != MessageType::BLOCK_REQUEST && req->id == requestId;
+    });
+    if (!msg) {
+        msg = findRefInList(unackMsgs_, [=](auto msg) {
+            return msg->type != MessageType::BLOCK_REQUEST && msg->requestId == requestId;
+        });
+        if (!msg) {
+            msg = findRefInList(blockResps_, [=](auto msg) {
+                return msg->requestId == requestId;
+            });
+        }
+    }
+    if (msg) {
+        clearMessage(msg);
+    }
+}
+
+int CoapChannel::addRequestHandler(const char* uri, coap_method method, coap_request_callback callback, void* callbackArg) {
+    if (!*uri || !callback) {
+        return SYSTEM_ERROR_INVALID_ARGUMENT;
+    }
+    if (std::strlen(uri) != 1) {
+        return SYSTEM_ERROR_NOT_SUPPORTED; // TODO: Support longer URIs
+    }
+    if (state_ != State::CLOSED) {
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    auto h = findInList(reqHandlers_, [=](auto h) {
+        return h->uri == *uri && h->method;
+    });
+    if (h) {
+        h->callback = callback;
+        h->callbackArg = callbackArg;
+    } else {
+        std::unique_ptr<RequestHandler> h(new(std::nothrow) RequestHandler(*uri, method, callback, callbackArg));
+        if (!h) {
+            return SYSTEM_ERROR_NO_MEMORY;
+        }
+        addToList(reqHandlers_, h.release());
+    }
+    return 0;
+}
+
+void CoapChannel::removeRequestHandler(const char* uri, coap_method method) {
+    if (std::strlen(uri) != 1) { // TODO: Support longer URIs
+        return;
+    }
+    if (state_ != State::CLOSED) {
+        LOG(ERROR, "Cannot remove handler while channel is open");
+        return;
+    }
+    auto h = findInList(reqHandlers_, [=](auto h) {
+        return h->uri == *uri && h->method == method;
+    });
+    if (h) {
+        removeFromList(reqHandlers_, h);
+        delete h;
+    }
+}
+
+int CoapChannel::addConnectionHandler(coap_connection_callback callback, void* callbackArg) {
+    if (!callback) {
+        return SYSTEM_ERROR_INVALID_ARGUMENT;
+    }
+    if (state_ != State::CLOSED) {
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    auto h = findInList(connHandlers_, [=](auto h) {
+        return h->callback == callback;
+    });
+    if (!h) {
+        std::unique_ptr<ConnectionHandler> h(new(std::nothrow) ConnectionHandler(callback, callbackArg));
+        if (!h) {
+            return SYSTEM_ERROR_NO_MEMORY;
+        }
+        addToList(connHandlers_, h.release());
```

Comment:
What's the advantage of using a unique_ptr here since you're immediately passing ownership to the list?

Reply:
I'm just being defensive here: if someone else adds another error condition before the ownership is passed to the list, they won't have a chance to leak the memory because it's already managed by a smart pointer.